File: solutions_2755486_0/java/bluepichu/wall.java
//Program by Matthew Savage
//With help from the TJHSST SCT Grader Example Class

import java.io.*;
import java.util.*;

public class wall{
	
	public static Long[] sortedKeys;
	
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new FileReader("C-small-attempt6.in"));
		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("wall.out")));
		
		int testCases = Integer.parseInt(br.readLine());
		
		for(int t = 1; t <= testCases; t++){
			pw.print("Case #" + t + ": ");
			HashSet[] arr = new HashSet[676061];
			int numTribes = Integer.parseInt(br.readLine());
			int last = 0;
			
			for(int i = 0; i < numTribes; i++){
				StringTokenizer st = new StringTokenizer(br.readLine(), " ");
				long d = Long.parseLong(st.nextToken());
				long n = Long.parseLong(st.nextToken());
				long w = Long.parseLong(st.nextToken());
				long e = Long.parseLong(st.nextToken());
				long s = Long.parseLong(st.nextToken());
				long dd = Long.parseLong(st.nextToken());
				long dp = Long.parseLong(st.nextToken());
				long ds = Long.parseLong(st.nextToken());
				
				for(int j = 0; j < n; j++){
					int day = (int)(d + dd * j);
					last = Math.max(last, day);
					long east = e + dp * j;
					long west = w + dp * j;
					long strength = s + ds * j;
					if(arr[day] == null){
						arr[day] = new HashSet<AttackData>();
					}
					arr[day].add(new AttackData(east, west, strength));
				}
			}
			
			HashSet<AttackData> successful = new HashSet<AttackData>();
			HashMap<Long, Long> wall = new HashMap<Long, Long>();
			sortedKeys = new Long[0];
			int ans = 0;
			
			for(int i = 0; i <= last+1; i++){
				if(arr[i] == null){
					continue;
				}
				for(AttackData attack : (HashSet<AttackData>)(arr[i])){
					if(success(attack, wall)){
						successful.add(attack);
					}
				}
				
				Iterator it = successful.iterator();
				
				while(it.hasNext()){
					ans++;
					AttackData ad = (AttackData)(it.next());
					it.remove();
					construct(ad, wall);
				}
			}
			pw.println(ans);
		}
		
		br.close();
		pw.close();
	}
	
	static class AttackData{
		public long east;
		public long west;
		public long strength;
		public AttackData(long e, long w, long s){
			east = e;
			west = w;
			strength = s;
		}
	}
	
	public static boolean success(AttackData ad, HashMap<Long, Long> wall){
		long h = 0;
		int i = 0;
		while(i < sortedKeys.length && sortedKeys[i] <= ad.west){
			h += wall.get(sortedKeys[i]);
			i++;
		}
		if(h < ad.strength){
			return true;
		}
		while(i < sortedKeys.length && sortedKeys[i] < ad.east){
			h += wall.get(sortedKeys[i]);
			
			if(h < ad.strength){
				return true;
			}
			
			i++;
		}
		return false;
	}
	
	public static void construct(AttackData ad, HashMap<Long, Long> wall){
		long h = 0;
		int i = 0;
		while(i < sortedKeys.length && sortedKeys[i] <= ad.west){
			h += wall.get(sortedKeys[i]);
			i++;
		}
		long added = 0;
		if(h < ad.strength){
			if(!wall.keySet().contains(ad.west)){
				wall.put(ad.west, (long)(0));
			}
			wall.put(ad.west, wall.get(ad.west) + ad.strength - h);
			added = ad.strength - h;
			h = ad.strength;
		}
		while(i < sortedKeys.length && sortedKeys[i] <= ad.east){
			if(wall.get(sortedKeys[i]) > 0){
				added -= wall.get(sortedKeys[i]);
				if(added >= 0){
					wall.put(sortedKeys[i], (long)(0));
				} else {
					wall.put(sortedKeys[i], -added);
					added = 0;
				}
			}
			h += wall.get(sortedKeys[i]);
			if(h < ad.strength){
				wall.put((long)sortedKeys[i], wall.get(sortedKeys[i]) + ad.strength - h);
				added += ad.strength - h;
			}
			
			i++;
		}
		if(!wall.keySet().contains(ad.east)){
			wall.put(ad.east, (long)(0));
		}
		wall.put(ad.east, wall.get(ad.east)-added);
		
		sortedKeys = wall.keySet().toArray(new Long[0]);
		Arrays.sort(sortedKeys);
	}
}

File: solutions_2755486_0/Python/Gautham/wall.py
import math

t = input()
for i in range(t):
	N = int(raw_input())
	d = [0]*N
	n = [0]*N
	w = [0]*N
	e = [0]*N
	s = [0]*N
	d_d = [0]*N
	d_p = [0]*N
	d_s = [0]*N
	
	dict = {}
	
	for k in range(N):
		d[k],n[k],w[k],e[k],s[k],d_d[k],d_p[k],d_s[k] = map(int,raw_input().split())
	success = 0
	day = 0
	t = N
	while (t>0):
		bdict = {}
		for k in range(N):
			tsucc = False
			if d[k] == day:
				for j in range(w[k],e[k]):
					if not j in dict:
						#print j,i,day
						tsucc = True
						if j in bdict:
							bdict[j] = max(bdict[j],s[k])
						else:
							bdict[j] = s[k]
					elif dict[j] < s[k]:
						#print j,i,day
						tsucc = True
						if j in bdict:
							bdict[j] = max(bdict[j],s[k])
						else:
							bdict[j] = s[k]
				if tsucc:
					success = success+1
				if(n[k]>1):
					n[k] = n[k]-1
					d[k] = d[k]+d_d[k]
					w[k] = w[k]+d_p[k]
					e[k] = e[k]+d_p[k]
					s[k] = s[k]+d_s[k]
				else:
					t = t-1
		for j in bdict:
			dict[j] = bdict[j]
		day = day+1
	print "Case #"+str(i+1)+": "+str(success)